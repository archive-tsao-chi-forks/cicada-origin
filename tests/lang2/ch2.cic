// 2. Doin' What Comes Naturally

{
  flavored_lentils : (String) -> Pair(String, String)
  flavored_lentils = (flavor) => cons(flavor, "lentils")

  garlic_flavored_lentils : Pair(String, String)
  garlic_flavored_lentils = flavored_lentils("garlic")

  id : (String) -> String
  id = (x) => x

  T : Type
  T = Pair(String, (String) -> String)

  id_in_pair : (x: String) -> T
  id_in_pair = (x) => cons(x, (x) => x)

  _ : T
  _ = id_in_pair("x")

  // The x occur in `(x) => x` should not be
  // effected by the substitution of outer `x`,
  // and `(x) => x` is equivalent to `(y) => y`,

  _ : Equal(T, id_in_pair("x"), cons("x", (y) => y))
  _ = same

  _
}
