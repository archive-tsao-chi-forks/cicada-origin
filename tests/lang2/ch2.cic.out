same: Equal((x2: String) * (x3: String) -> String, cons("x", (x3) => x3), cons("x", (x3) => x3))
